stdlib audit: use the "opened_channels" lock for channel refcounting

Octachron opened the pull request:
This PR implements a lock-based alternative to the atomic refcounting of channels in `runtime/io.c`.

Taking advantage of the fact that the refcounting is here to handle channel copies created by `caml_ml_out_channels_list`, this PR proposes to reacquire the lock for the list of opened channels when allocating copy and incrementing the refcounter of opened channels. With #11251 , this is enough to ensure that all updates to channel refcounters are protected by a lock.

Octachron committed:
stdlib audit: use lock for channel refcounting

Octachron committed:
stdlib channel: take lock in caml_alloc_chan

The `open_descriptor_in` function links channels to the list of the opened
channels. This makes it necessary for `caml_alloc_channel` to take the lock
when updating the reference counter to avoid a race with a concurrent
run of the `caml_ml_out_channels_list` function.

Another option would have been to initialize the reference counter to one in
`open_descriptor_in` before linking the channel. However, this change breaks
C-only channels which are using `caml_close_channel` to free the channel
resources when `refcount`=`0`. We could still make the change and identify
C-only channels with the `CHANNEL_FLAG_MANAGED_BY_GC` flag, but that would
still require to acquire a lock (but this time the channel lock).

Octachron commented:
I have pushed another commit that take care of  locking the `all_opened_channels` mutex in `caml_alloc_channel`, after adiscussion with @xavierleroy.

Briefly, the issue is that the channel is already linked in the list of all opened channels at the end of the call of `caml_open_descriptor_in`. We are thus subject to  a race with `caml_ml_openened_channel_lists` before the OCaml custom block is allocated. Taking the lock in `caml_alloc_channel` fixes this race.

Another option would be only link channels to the channel list once they are fully initialized, which removes one lock but splits the `caml_open_descriptor_{in,out}` functions in two variants (each): one for OCaml-managed channels and another for `C`-managed channels.

Another variant would be to change the channel refcount to count the number of references to all channels. However, this create issues with C-only channels that are currently using  `caml_close_channel` to free C resources when `channel->refcount`=0, and thus we will need to replace this access to `channel->refcount` by an access to `channel->flag &  CHANNEL_FLAG_MANAGED_BY_GC `.

Overall, taking the `all_opened_channels` lock twice for OCaml-managed channels sounds like a reasonable option.

xavierleroy reviewed (approved):
The changes look correct to me.  However, I think the code can be simplified further.  I also spotted a problem in `caml_finalize_channel` that was introduced  earlier.  See https://github.com/Octachron/ocaml/pull/7 .  I'm approving so that @Octachron can choose to merge the present PR and then I'll submit my changes separately.  But he can also choose to fuse the two PRs.

xavierleroy committed:
Revised handling of the list of all opened channels (#7)

- Don't put channels opened from C on this list, only those opened from OCaml
  and tracked by the GC.
- Simplify several functions accordingly.
- Fix an error in caml_finalize_channel where the channel could be
  unlinked from the list, then not freed because not flushed.

Octachron commented:
I have reviewed and merged @xavierleroy proposal which removes the C-managed channels from the list of opened channels (where there were ignored).

I am planning to merge soon.
